File: passwordtools.cpp
#include <iostream>
#include <string>
#include <cstdlib>
#include <ctime>
#include <cctype>
#include <cmath>

using namespace std;

//to generate random password
string generatePassword(int length, bool includeUppercase, bool includeDigits, bool includeSymbols) {
    const string lowercase = "abcdefghijklmnopqrstuvwxyz";
    const string uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    const string digits = "0123456789";
    const string symbols = "!@#$%^&*()_+[]{}|;:,.<>?";

    string characters = lowercase;
    if (includeUppercase) characters += uppercase;
    if (includeDigits) characters += digits;
    if (includeSymbols) characters += symbols;

    if (characters.empty()) {
        return "";
    }

    string password = "";
    for (int i = 0; i < length; ++i) {
        password += characters[rand() % characters.size()];
    }

    return password;
}

//to check strength
void checkPasswordStrength(const string& password) {
    bool hasUpper = false, hasLower = false, hasDigit = false, hasSymbol = false;
    for (char ch : password) {
        if (isupper(ch)) hasUpper = true;
        else if (islower(ch)) hasLower = true;
        else if (isdigit(ch)) hasDigit = true;
        else hasSymbol = true;
    }

    int length = password.length();
    cout << "Password strength: ";
    if (length >= 12 && hasUpper && hasLower && hasDigit && hasSymbol) {
        cout << "Strong" << endl;
    }
    else if (length >= 8 && ((hasUpper && hasLower && (hasDigit || hasSymbol)) || (hasLower && hasDigit && hasSymbol))) {
        cout << "Moderate" << endl;
    }
    else {
        cout << "Weak" << endl;
    }

    cout << "Recommendations: " << endl;
    if (length < 8) cout << "- Increase password length to at least 8 characters." << endl;
    if (!hasUpper) cout << "- Add uppercase letters." << endl;
    if (!hasLower) cout << "- Add lowercase letters." << endl;
    if (!hasDigit) cout << "- Include digits." << endl;
    if (!hasSymbol) cout << "- Include symbols." << endl;
}

//to estimate brute-force time
void estimateBruteForceTime(const string& password) {
    int length = password.length();
    bool hasUpper = false, hasLower = false, hasDigit = false, hasSymbol = false;

    for (char ch : password) {
        if (isupper(ch)) hasUpper = true;
        else if (islower(ch)) hasLower = true;
        else if (isdigit(ch)) hasDigit = true;
        else hasSymbol = true;
    }

    int characterSetSize = 0;
    if (hasLower) characterSetSize += 26;
    if (hasUpper) characterSetSize += 26;
    if (hasDigit) characterSetSize += 10;
    if (hasSymbol) characterSetSize += 32; // Approximate number of common symbols

    // Total possible combinations
    unsigned long long totalCombinations = pow(characterSetSize, length);

    // Assuming a hypothetical cracking speed
    const unsigned long long attemptsPerSecond = 1000000; // 1 Million attempts per second

    // Time to crack in seconds
    unsigned long long seconds = totalCombinations / attemptsPerSecond;

    // Convert time to larger units
    unsigned long long minutes = seconds / 60;
    unsigned long long hours = minutes / 60;
    unsigned long long days = hours / 24;
    unsigned long long years = days / 365;

    cout << "Estimated brute-force cracking time: ";
    if (years > 0) {
        cout << years << " years" << endl;
    }
    else if (days > 0) {
        cout << days << " days" << endl;
    }
    else if (hours > 0) {
        cout << hours << " hours" << endl;
    }
    else if (minutes > 0) {
        cout << minutes << " minutes" << endl;
    }
    else {
        cout << seconds << " seconds" << endl;
    }
}

int main() {
    // Seed random number generator
    srand(static_cast<unsigned int>(time(nullptr)));

    char continueChoice;

    do {
        int choice;
        cout << "Choose an option:" << endl;
        cout << "1. Generate a new password" << endl;
        cout << "2. Check the strength of an existing password" << endl;
        cout << "3. Estimate brute-force cracking time for a password" << endl;
        cout << "Enter your choice (1, 2, or 3): ";
        cin >> choice;

        if (choice == 1) {
            int length;
            bool includeUppercase, includeDigits, includeSymbols;

            // Input from the user
            cout << "Enter password length: ";
            cin >> length;

            cout << "Include uppercase letters? (1 for Yes, 0 for No): ";
            cin >> includeUppercase;

            cout << "Include digits? (1 for Yes, 0 for No): ";
            cin >> includeDigits;

            cout << "Include symbols? (1 for Yes, 0 for No): ";
            cin >> includeSymbols;

            // Generate and display the password
            string password = generatePassword(length, includeUppercase, includeDigits, includeSymbols);
            cout << "Generated Password: " << password << endl;

            // Check password strength
            checkPasswordStrength(password);

            // Estimate brute-force time
            estimateBruteForceTime(password);
        }
        else if (choice == 2) {
            string password;
            cout << "Enter the password to check: ";
            cin >> password;

            // Check password strength
            checkPasswordStrength(password);
        }
        else if (choice == 3) {
            string password;
            cout << "Enter the password to estimate brute-force time: ";
            cin >> password;

            // Estimate brute-force time
            estimateBruteForceTime(password);
        }
        else {
            cout << "Invalid choice. Please select 1, 2, or 3." << endl;
        }

        cout << "Do you want to perform another operation? (Y/N): ";
        cin >> continueChoice;

    } while (tolower(continueChoice) == 'y');

    cout << "Thank you for using the program. Goodbye!" << endl;

    return 0;
}

File: passwordtools.py
import random
import string
import math

def generate_password(length, include_uppercase, include_digits, include_symbols):
    lowercase = string.ascii_lowercase
    uppercase = string.ascii_uppercase
    digits = string.digits
    symbols = string.punctuation

    characters = lowercase
    if include_uppercase:
        characters += uppercase
    if include_digits:
        characters += digits
    if include_symbols:
        characters += symbols

    if not characters:
        return ""

    return ''.join(random.choice(characters) for _ in range(length))

def check_password_strength(password):
    has_upper = any(c.isupper() for c in password)
    has_lower = any(c.islower() for c in password)
    has_digit = any(c.isdigit() for c in password)
    has_symbol = any(c in string.punctuation for c in password)
    length = len(password)

    print("Password strength: ", end="")
    if length >= 12 and has_upper and has_lower and has_digit and has_symbol:
        print("Strong")
    elif length >= 8 and ((has_upper and has_lower and (has_digit or has_symbol)) or (has_lower and has_digit and has_symbol)):
        print("Moderate")
    else:
        print("Weak")

    print("Recommendations:")
    if length < 8:
        print("- Increase password length to at least 8 characters.")
    if not has_upper:
        print("- Add uppercase letters.")
    if not has_lower:
        print("- Add lowercase letters.")
    if not has_digit:
        print("- Include digits.")
    if not has_symbol:
        print("- Include symbols.")

def estimate_bruteforce_time(password):
    length = len(password)
    has_upper = any(c.isupper() for c in password)
    has_lower = any(c.islower() for c in password)
    has_digit = any(c.isdigit() for c in password)
    has_symbol = any(c in string.punctuation for c in password)

    charset_size = 0
    if has_lower:
        charset_size += 26
    if has_upper:
        charset_size += 26
    if has_digit:
        charset_size += 10
    if has_symbol:
        charset_size += 32

    total_combinations = math.pow(charset_size, length)
    attempts_per_second = 1_000_000_000  # 1 billion attempts per second
    seconds = total_combinations / attempts_per_second

    years = seconds / (60 * 60 * 24 * 365)
    if years > 1:
        print(f"Estimated brute-force cracking time: {years:.2f} years")
    elif seconds > 86400:
        print(f"Estimated brute-force cracking time: {seconds / 86400:.2f} days")
    elif seconds > 3600:
        print(f"Estimated brute-force cracking time: {seconds / 3600:.2f} hours")
    elif seconds > 60:
        print(f"Estimated brute-force cracking time: {seconds / 60:.2f} minutes")
    else:
        print(f"Estimated brute-force cracking time: {seconds:.2f} seconds")


def main():
    while True:
        print("Choose an option:")
        print("1. Generate a new password")
        print("2. Check the strength of an existing password")
        print("3. Estimate brute-force cracking time for a password")
        print("4. Exit")

        choice = input("Enter your choice (1, 2, 3, or 4): ")

        if choice == '1':
            length = int(input("Enter password length: "))
            include_uppercase = input("Include uppercase letters? (y/n): ").lower() == 'y'
            include_digits = input("Include digits? (y/n): ").lower() == 'y'
            include_symbols = input("Include symbols? (y/n): ").lower() == 'y'
            password = generate_password(length, include_uppercase, include_digits, include_symbols)
            print(f"Generated Password: {password}")
            check_password_strength(password)
            estimate_bruteforce_time(password)

        elif choice == '2':
            password = input("Enter the password to check: ")
            check_password_strength(password)

        elif choice == '3':
            password = input("Enter the password to estimate brute-force time: ")
            estimate_bruteforce_time(password)

        elif choice == '4':
            print("Thank you for using the program. Goodbye!")
            break

        else:
            print("Invalid choice. Please select 1, 2, 3, or 4.")

if __name__ == "__main__":
    main()
